File: 16threeSumClosest/threeSumClosest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        result = sum(itertools.islice(nums, 3))
        index = 0
        while index != len(nums) - 2:
            num = nums[index]
            left, right = index + 1, len(nums) - 1
            while left < right:
                numsLeft, numsRight = map(nums.__getitem__, (left, right))
                total = num + numsLeft + numsRight
                if total == target: return target 
                if total <= target:
                    while left < right and nums[left] == numsLeft: left += 1
                if total >= target:
                    while left < right and nums[right] == numsRight: right -= 1                             
                if abs(target-total) < abs(result - target): result = total           
            while index != len(nums) - 2 and nums[index] == num: index += 1
        return result

File: 16threeSumClosest/threeSumClosest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        var result = Arrays.stream(nums).limit(3).sum();
        for (var index = 0; index != nums.length - 2;)
        {
            final var num = nums[index];
            var left = index + 1;
            var right = nums.length - 1;
            while (left != right)
            {
                final var numsLeft = nums[left];
                final var numsRight = nums[right];
                final var total = num + numsLeft + numsRight;
                if (total == target) return target;
                if (total <= target) while (left != right && numsLeft == nums[left]) ++left;
                if (total >= target) while (left != right && numsRight == nums[right]) --right;
                if (Math.abs(total - target) < Math.abs(result - target)) result = total;
            }
            while (nums[index] == num && index != nums.length - 2) ++index;
        }
        return result;
    }
}
